Feature/ws restore

mebassett opened the pull request:
adds websocket server to handle legacy restore operations
also adds external config
also does some eslinting.

Commit:
add - websocket server to restore daedalus addresses

this is mostly a verbatim version of what was in the old backend.
There isn't that much work, two things took me awhile:
 (1) tracing down the old tangata repo (I knew I had seen it before but couldn't remember it)
     to find out what is going on in the `utxos` table referenced in old code.
 (2) looking at yoroi-frontend code to figure out how this is being used.

Those two things weren't strictly necessary to do this (well, (1) was useful for figuring
out the graphql query, but even so, I'm not 100% sure its correct).

There are no tests, and I have not tested to make sure that this works (I'm not sure how?)
so I am relying on code reviewers to let me know.

Commit:
add - external config file.

we're using `config.get('string for json path to param')`, which I do not love.
but I am not sure how typescript can get config object types.

I'd love to do `config.server.graphqlEndpoint`, but I couldn't figure out how.

See https://stackoverflow.com/questions/50813591/how-to-use-node-config-in-typescript

Commit:
add - eslint

Commit:
eslinting

Commit:
eslint after rebase

SebastienGllmt commented on src/ws-server.ts:
```diff
@@ -0,0 +1,53 @@
+*/
+const GRAPHQL_QUERY = `
+{
+  utxos(distinct_on: address) {
```
Strictly speaking, we need all UTXOs where the address is a Daedalus address.

What is a Daedalus address you ask?

Cardano (starting in Shelley) has multiple address types:
- Base
- Enterprise
- Pointer
- Bootstrap

Bootstrap refers to the type of addresses used in the Byron era. In Yoroi, these addresses start with Ae2 and in Daedalus, these addresses start with Dd

Although this query will not fail if we return more addresses than needed, it's just less efficient (so don't spend too much time on this if it doesn't look easy to do). So preferably, we would only return bootstrap addresses that are of the Daedalus variety.
You can tell something is a bootstrap address if it starts with the bits `0b1000` (hopefully GraphQL provides at least this much). I don't suggest doing something like "starting starts with Dd" because the Dd prefix is network-specific (by design so that addresses look different on the testnet)

Sorry this is maybe a complex explanation. Let me know if you have any questions.

mebassett replied:
you're right that this query is _slow_.  though it was the same process (filtering out addresses after you got it) as done in the old backend.  

I am confident I can do a graphql filter for strings starting with `Ae2` or `Dd` but I don't think I can check if it starts with a certain sequence of bytes.  the underlying sql tables and views hold addresses as `varchar`.  there's probably a useful postgresql function for the filter, but its not in graphql.

mebassett replied:
this doesn't actually work.  the version of cardano-graphql we are using in production does not support `distinct_on: address` in a `utxos` query.  a more recent version does support it, but it breaks other things.  since we should be moving to the top of upstream anyway, I'll fix it on upstream and then come back to this.

mebassett replied:
alas this does not work. Even balas with the PR to cardano-graphql.

If I run the following sql query on the schema (views) created by cardano-graphql I get lots of valid addresses:
`select distinct address from  "Utxo" where address like 'Ae2%';`

yet that same graphql query returns a completely different, smaller resultset.  notably, it lacks anything that looks like a bootstrap address.  everything in the graphql resultset starts with `2w1sdSJu3GVd`

mebassett replied:
ah, I see the problem.  the query works fine if you do it on hasura directly, but the cardano-graphql node app actually truncates the query.  hmm.  tough nuggets.  because the result set is so big it might be best to do this in sql directly afterall.

Commit:
chore - updating code so it works with latest cardano-graphql upstream.

closes https://github.com/Emurgo/yoroi-graphql-migration-backend/issues/10

Commit:
bugfix - need to catch errors. wip wip.

Commit:
giving up on graphql and using sql.

mebassett commented:
so the node app that acts as the frontend for cardano-graphql limits the size of the resultset, so you can't really use it for this.  you can use hasura directly, or just sql.  I opted for sql, since we already do that.  Its a bit of a hassle getting things ready to ping hasura directly.  would need to change the docker-compose file in production.  

this introduces websockets, and in production (at least on the aws server I have access to) this app is behind an nginx proxy.  so ningx will also need to proxy websocket connections, something like
```
    location / { 
        proxy_pass http://localhost:8082;
        proxy_http_version 1.1;
        proxy_set_header Upgrade $http_upgrade;
        proxy_set_header Connection "Upgrade";
        proxy_set_header Host $host;
    } 
```

Commit:
fix - bestblock for cardano-graphql#2.0.0-beta.0.

Commit:
Update tx history endpoint for Shelley txs [ch5527]

After about a half day of looking, I cannot find `ttl` anwhere in the
schema created by cardano-db-schema.  On Seba's advice, I have just used
the maxint.
`ttl` should be a slot_no, and these are stored in postgresql as a `uinteger`
which is just an `integer` ie `int4`  whose value >0.  So took the largest int4.

Certificates are not implemented.

Commit:
mod - removing ttl.

Commit:
add - block era type to tx history obj.

Commit:
whoops, last commit was daft

Commit:
fix - txHistory empty withdrawals should be an empty list, not null.

why didn't typescript yell at me for this mistake? hmmm.

Commit:
mod - txHistory withdrawals should never be null, but may be []

Commit:
Merge branch 'master' into feature/ws-restore

Commit:
del - removing txBodies endpoint.
